raw-mutex-traits/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+#![no_std]
+
+use core::{cell::UnsafeCell, sync::atomic::{AtomicBool, Ordering}};
+
+/// Raw mutex trait.
+///
+/// This mutex is "raw", which means it does not actually contain the protected data, it
+/// just implements the mutex mechanism. For most uses you should use [`super::Mutex`] instead,
+/// which is generic over a RawMutex and contains the protected data.
+///
+/// Note that, unlike other mutexes, implementations only guarantee no
+/// concurrent access from other threads: concurrent access from the current
+/// thread is allowed. For example, it's possible to lock the same mutex multiple times reentrantly.
+///
+/// Therefore, locking a `RawMutex` is only enough to guarantee safe shared (`&`) access
+/// to the data, it is not enough to guarantee exclusive (`&mut`) access.
+///
+/// # Safety
+///
+/// RawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently.
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+pub unsafe trait RawMutex {
+    /// Create a new `RawMutex` instance.
+    ///
+    /// This is a const instead of a method to allow creating instances in const context.
+    const INIT: Self;
+
+    /// Lock this `RawMutex`.
+    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
+}
+
+/// Blocking mutex (not async)
+///
+/// Provides a blocking mutual exclusion primitive backed by an implementation of [`raw::RawMutex`].
+///
+/// Which implementation you select depends on the context in which you're using the mutex, and you can choose which kind
+/// of interior mutability fits your use case.
+///
+/// Use [`CriticalSectionMutex`] when data can be shared between threads and interrupts.
+///
+/// Use [`NoopMutex`] when data is only shared between tasks running on the same executor.
+///
+/// Use [`ThreadModeMutex`] when data is shared between tasks running on the same executor but you want a global singleton.
+///
+/// In all cases, the blocking mutex is intended to be short lived and not held across await points.
+/// Use the async [`Mutex`](crate::mutex::Mutex) if you need a lock that is held across await points.
+pub struct BlockingMutex<R, T: ?Sized> {
```

Comment:
perhaps we can do some bikeshedding on the name
- crate should be raw-mutex-trait*s* or raw-mutex-trait?  there's only one trait.
- perhaps "BlockingMutex" only makes sense in the context of embsasy-sync, and it could be just "Mutex" here?

Reply:
I think making it clear that it is a blocking mutex is a good thing, based on embassy sync and maitake sync, which will each define their own async `Mutex` type.

Basically, I want to save the shorter `Mutex` name for the types we want USERS to reach for, while `BlockingMutex` is mostly for people building other things on top of it.

Reply:
re: crate name: yeah, idk. I didn't want to be `embassy-sync-stable-parts` or something, since I want to make it clear that the hope for this is to be useful to other non-embassy libraries, like maitake-sync. Bikeshedding welcome on the crate name tho.

Reply:
@Dirbaio I went with `scoped-mutex` as a name, since it fits the reality better!